### PR TITLE
Add missing docs for System.Collections.Immutable

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
## Summary

Contributes to https://github.com/dotnet/dotnet-api-docs/issues/3914

### CODE REVIEW

File: xml/System.Collections.Immutable/ImmutableArray.xml [hunk 1]
```diff
@@ -679,12 +679,12 @@ This overload allows helper methods or custom builder classes to efficiently avo
         <Parameter Name="selector" Type="System.Func&lt;TSource,TResult&gt;" />
       </Parameters>
       <Docs>
-        <typeparam name="TSource">To be added.</typeparam>
-        <typeparam name="TResult">To be added.</typeparam>
+        <typeparam name="TSource">The type of element stored in the source array.</typeparam>
+        <typeparam name="TResult">The type of element to be stored in the target array.</typeparam>
```
Comment (carlossanlop) [inline xml/System.Collections.Immutable/ImmutableArray.xml:683]:
```suggestion
        <typeparam name="TResult">The type of element to store in the target array.</typeparam>
```

File: xml/System.Collections.Immutable/ImmutableArray.xml [hunk 2]
```diff
@@ -728,14 +728,14 @@ This overload allows helper methods or custom builder classes to efficiently avo
...
```
Comment (carlossanlop) [inline xml/System.Collections.Immutable/ImmutableArray.xml:732]: I would say the same language should be applied on all similar APIs:
```suggestion
        <typeparam name="TResult">The type of element to store in the target array.</typeparam>
```

File: xml/System.Collections.Immutable/ImmutableArray.xml [hunk 3]
```diff
@@ -779,14 +779,14 @@ This overload allows helper methods or custom builder classes to efficiently avo
         <Parameter Name="arg" Type="TArg" />
       </Parameters>
       <Docs>
-        <typeparam name="TSource">To be added.</typeparam>
-        <typeparam name="TArg">To be added.</typeparam>
-        <typeparam name="TResult">To be added.</typeparam>
+        <typeparam name="TSource">The type of element stored in the source array.</typeparam>
+        <typeparam name="TArg">The type of argument to be passed to the selector mapping function.</typeparam>
+        <typeparam name="TResult">The type of element to be stored in the target array.</typeparam>
```
Comment (carlossanlop) [inline xml/System.Collections.Immutable/ImmutableArray.xml:784]:
```suggestion
        <typeparam name="TResult">The type of element to store in the target array.</typeparam>
```

File: xml/System.Collections.Immutable/ImmutableArray.xml [hunk 4]
```diff
@@ -779,14 +779,14 @@ This overload allows helper methods or custom builder classes to efficiently avo
         <Parameter Name="arg" Type="TArg" />
       </Parameters>
       <Docs>
-        <typeparam name="TSource">To be added.</typeparam>
-        <typeparam name="TArg">To be added.</typeparam>
-        <typeparam name="TResult">To be added.</typeparam>
+        <typeparam name="TSource">The type of element stored in the source array.</typeparam>
+        <typeparam name="TArg">The type of argument to be passed to the selector mapping function.</typeparam>
```
Comment (carlossanlop) [inline xml/System.Collections.Immutable/ImmutableArray.xml:783]:
```suggestion
        <typeparam name="TArg">The type of argument to pass to the selector mapping function.</typeparam>
```

File: xml/System.Collections.Immutable/ImmutableArray`1+Builder.xml [hunk 5]
```diff
@@ -461,7 +461,15 @@
       <Parameters />
       <Docs>
         <summary>Removes all items from the array.</summary>
-        <remarks>To be added.</remarks>
+        <remarks>
+          <format type="text/markdown"><![CDATA[
+
+## Remarks
+
+<xref:System.Collections.Generic.ICollection%601.Count%2A> must be set to 0, and references to other objects from elements of the collection must be released.
```
Comment (carlossanlop) [inline xml/System.Collections.Immutable/ImmutableArray`1+Builder.xml:469]: Is this remark relevant to the user? Sounds like implementation details that for some reason were added to the remarks in the triple slash comments.

File: xml/System.Collections.Immutable/ImmutableArray`1+Builder.xml [hunk 6]
```diff
@@ -611,7 +627,25 @@
...
+ Default implementations of collections in the <xref:System.Collections.Generic?displayProperty=nameWithType> namespace aren't synchronized.   
+ 
+ 
+ ## Examples   The following example demonstrates how to implement the <xref:System.Collections.Generic.IEnumerable%601> interface and uses that implementation to create a LINQ query. When you implement <xref:System.Collections.Generic.IEnumerable%601>, you must also implement <xref:System.Collections.Generic.IEnumerator%601> or, for C# only, you can use the [yield](~/docs/csharp/language-reference/keywords/yield.md) keyword. Implementing <xref:System.Collections.Generic.IEnumerator%601> also requires <xref:System.IDisposable> to be implemented, which you will see in this example.
```
Comment (carlossanlop) [inline xml/System.Collections.Immutable/ImmutableArray`1+Builder.xml:643]: This is a bug I should fix in the tool - If an `## Examples` line is detected, a new line should be appended at the end:
```suggestion
 ## Examples
 
 The following example demonstrates how to implement the <xref:System.Collections.Generic.IEnumerable%601> interface and uses that implementation to create a LINQ query. When you implement <xref:System.Collections.Generic.IEnumerable%601>, you must also implement <xref:System.Collections.Generic.IEnumerator%601> or, for C# only, you can use the [yield](~/docs/csharp/language-reference/keywords/yield.md) keyword. Implementing <xref:System.Collections.Generic.IEnumerator%601> also requires <xref:System.IDisposable> to be implemented, which you will see in this example.   
```

File: xml/System.Collections.Immutable/ImmutableArray`1+Builder.xml [hunk 7]
```diff
@@ -611,7 +627,25 @@
...
+ Default implementations of collections in the <xref:System.Collections.Generic?displayProperty=nameWithType> namespace aren't synchronized.   
+ 
+ 
+ ## Examples   The following example demonstrates how to implement the <xref:System.Collections.Generic.IEnumerable%601> interface and uses that implementation to create a LINQ query. When you implement <xref:System.Collections.Generic.IEnumerable%601>, you must also implement <xref:System.Collections.Generic.IEnumerator%601> or, for C# only, you can use the [yield](~/docs/csharp/language-reference/keywords/yield.md) keyword. Implementing <xref:System.Collections.Generic.IEnumerator%601> also requires <xref:System.IDisposable> to be implemented, which you will see in this example.   
+ [!code-csharp[System.Collections.Generic.IEnumerableEx#1](~/samples/snippets/csharp/VS_Snippets_CLR_System/system.collections.generic.ienumerableex/cs/program.cs#1)] [!code-vb[System.Collections.Generic.IEnumerableEx#1](~/samples/snippets/visualbasic/VS_Snippets_CLR_System/system.collections.generic.ienumerableex/vb/module1.vb#1)]
```
Comment (carlossanlop) [inline xml/System.Collections.Immutable/ImmutableArray`1+Builder.xml:644]: Unless the examples are explicitly implementing the current API, then they should be removed from the remarks.
Naturally the tool is not smart enough to do this.

File: xml/System.Collections.Immutable/ImmutableArray`1+Builder.xml [hunk 8]
```diff
@@ -646,7 +680,15 @@
         <param name="item">The item to locate in the array.</param>
         <summary>Determines the index of a specific item in the array.</summary>
         <returns>The index of <paramref name="item" /> if it's found in the list; otherwise, -1.</returns>
-        <remarks>To be added.</remarks>
+        <remarks>
+          <format type="text/markdown"><![CDATA[
+
+## Remarks
+
+If an object occurs multiple times in the list, the <xref:System.Collections.Generic.IList%601.IndexOf%2A> method always returns the first instance found.
```
Comment (carlossanlop) [inline xml/System.Collections.Immutable/ImmutableArray`1+Builder.xml:688]: This remark was imported from the interface method implemented by this method. Is it still relevant?

File: xml/System.Collections.Immutable/ImmutableArray`1.xml [hunk 9]
```diff
@@ -325,9 +325,9 @@
       </TypeParameters>
       <Parameters />
       <Docs>
-        <typeparam name="TOther">To be added.</typeparam>
+        <typeparam name="TOther">The type of array element to be returned.</typeparam>
```
Comment (carlossanlop) [inline xml/System.Collections.Immutable/ImmutableArray`1.xml:328]:
```suggestion
        <typeparam name="TOther">The type of array element to return.</typeparam>
```

File: xml/System.Collections.Immutable/ImmutableArray`1.xml [hunk 10]
```diff
@@ -325,9 +325,9 @@
...
         <summary>Initializes a new instance of the <see cref="T:System.Collections.Immutable.ImmutableArray`1" /> struct by casting the underlying array to an array of type <typeparamref name="TOther" />.</summary>
-        <returns>To be added.</returns>
+        <returns>An ImmutableArray instance with elements cast to the new type.</returns>
```
Comment (carlossanlop) [inline xml/System.Collections.Immutable/ImmutableArray`1.xml:330]: Either use a see cref, or use natural language. Usually, natural language is preferred, especially if we are already in the context where the user knows what API or type you're talking about:
```suggestion
        <returns>An immutable array instance with elements cast to the new type.</returns>
```

File: xml/System.Collections.Immutable/ImmutableArray`1.xml [hunk 11]
```diff
@@ -366,10 +366,10 @@
         <Parameter Name="items" Type="System.Collections.Immutable.ImmutableArray&lt;TDerived&gt;" />
       </Parameters>
       <Docs>
-        <typeparam name="TDerived">To be added.</typeparam>
+        <typeparam name="TDerived">The type of array element to be returned.</typeparam>
```
Comment (carlossanlop) [inline xml/System.Collections.Immutable/ImmutableArray`1.xml:369]:
```suggestion
        <typeparam name="TDerived">The type of array element to return.</typeparam>
```

File: xml/System.Collections.Immutable/ImmutableArray`1.xml [hunk 12]
```diff
@@ -366,10 +366,10 @@
...
         <param name="items">The array to initialize the array with. No copy is made.</param>
-        <summary>Initializes a new instance of the <see cref="T:System.Collections.Immutable.ImmutableArray`1" /> struct based on the contents             of an existing instance, allowing a covariant static cast to efficiently reuse the existing array.</summary>
-        <returns>To be added.</returns>
+        <summary>Initializes a new instance of the <see cref="T:System.Collections.Immutable.ImmutableArray`1" /> struct based on the contents of an existing instance, allowing a covariant static cast to efficiently reuse the existing array.</summary>
+        <returns>An ImmutableArray instance with elements cast to the new type.</returns>
```
Comment (carlossanlop) [inline xml/System.Collections.Immutable/ImmutableArray`1.xml:372]:
```suggestion
        <returns>An immutable array instance with elements cast to the new type.</returns>
```

File: xml/System.Collections.Immutable/ImmutableArray`1.xml [hunk 13]
```diff
@@ -609,7 +617,29 @@
         <summary>Indicates whether specified array is equal to this array.</summary>
         <returns>
           <see langword="true" /> if <paramref name="other" /> is equal to this array; otherwise, <see langword="false" />.</returns>
-        <remarks>To be added.</remarks>
+        <remarks>
+          <format type="text/markdown"><![CDATA[
+
+## Remarks
```
Comment (carlossanlop) [inline xml/System.Collections.Immutable/ImmutableArray`1.xml:623]: Do these remarks apply for this API too?

File: xml/System.Collections.Immutable/ImmutableArray`1.xml [hunk 14]
```diff
@@ -735,7 +765,15 @@
         <param name="item">The item to search for.</param>
         <summary>Searches the array for the specified item.</summary>
         <returns>The zero-based index position of the item if it is found, or -1 if it is not.</returns>
-        <remarks>To be added.</remarks>
+        <remarks>
+          <format type="text/markdown"><![CDATA[
+
+## Remarks
+
+If an object occurs multiple times in the list, the <xref:System.Collections.Generic.IList%601.IndexOf%2A> method always returns the first instance found.
```
Comment (carlossanlop) [inline xml/System.Collections.Immutable/ImmutableArray`1.xml:773]: This sounds like it's implementation specific. Does this API do the same?

File: xml/System.Collections.Immutable/ImmutableArray`1.xml [hunk 15]
```diff
@@ -2421,8 +2484,8 @@
         <ReturnType>System.Int32</ReturnType>
       </ReturnValue>
       <Docs>
-        <summary>Gets the number of array in the collection.</summary>
-        <value>To be added.</value>
+        <summary>Gets the number of items in the collection.</summary>
+        <value>Number of items in the collection.</value>
```
Comment (carlossanlop) [inline xml/System.Collections.Immutable/ImmutableArray`1.xml:2488]:
```suggestion
        <value>The number of items in the collection.</value>
```

File: xml/System.Collections.Immutable/ImmutableArray`1.xml [hunk 16]
```diff
@@ -2540,7 +2603,7 @@
       </ReturnValue>
       <Docs>
         <summary>Gets the size of the array.</summary>
-        <value>To be added.</value>
+        <value>Number of items in the collection.</value>
```
Comment (carlossanlop) [inline xml/System.Collections.Immutable/ImmutableArray`1.xml:2606]:
```suggestion
        <value>The number of items in the collection.</value>
```

File: xml/System.Collections.Immutable/ImmutableArray`1.xml [hunk 17]
```diff
@@ -2576,9 +2639,25 @@
...
+ Enumerating through a collection is intrinsically not a thread-safe procedure. Even when a collection is synchronized, other threads can still modify the collection, which causes the enumerator to throw an exception. To guarantee thread safety during enumeration, you can either lock the collection during the entire enumeration or catch the exceptions resulting from changes made by other threads.   
+ The following code example shows how to lock the collection using the <xref:System.Collections.ICollection.SyncRoot%2A> property during the entire enumeration.   
+ [!code-cpp[System.Collections.ICollection#1](~/samples/snippets/cpp/VS_Snippets_CLR_System/system.collections.icollection/cpp/remarks.cpp#1)] [!code-csharp[System.Collections.ICollection#1](~/samples/snippets/csharp/VS_Snippets_CLR_System/system.collections.icollection/cs/remarks.cs#1)] [!code-vb[System.Collections.ICollection#1](~/samples/snippets/visualbasic/VS_Snippets_CLR_System/system.collections.icollection/vb/remarks.vb#1)]
```
Comment (carlossanlop) [inline xml/System.Collections.Immutable/ImmutableArray`1.xml:2657]: These examples are probably not for this API specificaly, so they should be removed.

File: xml/System.Collections.Immutable/ImmutableDictionary`2+Builder.xml [hunk 18]
```diff
@@ -482,7 +527,25 @@
...
+ The C# language uses the [this](~/docs/csharp/language-reference/keywords/this.md) keyword to define the indexers instead of implementing the <xref:System.Collections.Generic.IDictionary%602.Item%2A> property. Visual Basic implements <xref:System.Collections.Generic.IDictionary%602.Item%2A> as a default property, which provides the same indexing functionality.   
+ Implementations can vary in whether they allow `key` to be `null`.   
+ 
+ 
+ ## Examples   The following code example uses the <xref:System.Collections.Generic.IDictionary%602.Item%2A> property (the indexer in C#) to retrieve values, demonstrating that a <xref:System.Collections.Generic.KeyNotFoundException> is thrown when a requested key is not present, and showing that the value associated with a key can be replaced.
```
Comment (carlossanlop) [inline xml/System.Collections.Immutable/ImmutableDictionary`2+Builder.xml:542]: If the examples do not apply for this API, please remove them. Otherwise:
```suggestion
 ## Examples
 
 The following code example uses the <xref:System.Collections.Generic.IDictionary%602.Item%2A> property (the indexer in C#) to retrieve values, demonstrating that a <xref:System.Collections.Generic.KeyNotFoundException> is thrown when a requested key is not present, and showing that the value associated with a key can be replaced.   
```

File: xml/System.Collections.Immutable/ImmutableDictionary`2+Builder.xml [hunk 19]
```diff
@@ -693,9 +786,9 @@
         <Parameter Name="arrayIndex" Type="System.Int32" />
       </Parameters>
       <Docs>
-        <param name="array">To be added.</param>
-        <param name="arrayIndex">To be added.</param>
-        <summary>To be added.</summary>
+        <param name="array">The one-dimensional array of type <see cref="T:System.Collections.Generic.KeyValuePair`2" /> that is the destination of the elements copied from the dictionary. The array must have zero-based indexing.</param>
```
Comment (carlossanlop) [inline xml/System.Collections.Immutable/ImmutableDictionary`2+Builder.xml:789]: The type already shows up in MS Docs next to the argument, no need to mention it again.
```suggestion
        <param name="array">The one-dimensional array that is the destination of the elements copied from the dictionary. The array must have zero-based indexing.</param>
```

File: xml/System.Collections.Immutable/ImmutableDictionary`2+Builder.xml [hunk 20]
```diff
@@ -822,7 +949,25 @@
...
+ Default implementations of collections in the <xref:System.Collections.Generic?displayProperty=nameWithType> namespace aren't synchronized.   
+ 
+ 
+ ## Examples   The following example demonstrates how to implement the <xref:System.Collections.Generic.IEnumerable%601> interface and uses that implementation to create a LINQ query. When you implement <xref:System.Collections.Generic.IEnumerable%601>, you must also implement <xref:System.Collections.Generic.IEnumerator%601> or, for C# only, you can use the [yield](~/docs/csharp/language-reference/keywords/yield.md) keyword. Implementing <xref:System.Collections.Generic.IEnumerator%601> also requires <xref:System.IDisposable> to be implemented, which you will see in this example.
```
Comment (carlossanlop) [inline xml/System.Collections.Immutable/ImmutableDictionary`2+Builder.xml:965]: Same as before - keep if the examples apply:
```suggestion
 ## Examples
 
 The following example demonstrates how to implement the <xref:System.Collections.Generic.IEnumerable%601> interface and uses that implementation to create a LINQ query. When you implement <xref:System.Collections.Generic.IEnumerable%601>, you must also implement <xref:System.Collections.Generic.IEnumerator%601> or, for C# only, you can use the [yield](~/docs/csharp/language-reference/keywords/yield.md) keyword. Implementing <xref:System.Collections.Generic.IEnumerator%601> also requires <xref:System.IDisposable> to be implemented, which you will see in this example.   
```

File: xml/System.Collections.Immutable/ImmutableArray`1+Builder.xml (hunk 5 above)
Comment (eiriktsarpalis) [reply xml/System.Collections.Immutable/ImmutableArray`1+Builder.xml:469]: Agreed, will remove.

**Comment by BillWagner** [comment]
@carlossanlop It looks like all changes have been addressed.

**Comment by eiriktsarpalis** [comment]
@carlossanlop this should be ready to go in.
